# import library

In [1]:
import pandas as pd
import numpy as np
import os

# import data and define file path

In [3]:
path = "C:/Users/hardi/Documents/Instacart_Baskey_analysis"

In [4]:
df_pro = pd.read_csv(os.path.join(path, "Data", "original_data","products.csv"),index_col=False)

In [6]:
df_ords = pd.read_csv(os.path.join(path,"Data", "prepare_data", "orders_wrangled.csv"))

## mixed type data 

In [7]:
## create empty datafarme

In [8]:
df_test = pd.DataFrame()

In [9]:
df_test['mix'] = ['a', 'b', 1, True]

In [21]:
## check if any mix tytpe data present in columns
for col in df_test.columns.to_list():
    
    werid = (df_test[col].map(type) != df_test[col].iloc[0].__class__).any()
    if werid:
        print(col)

mix


In [22]:
#convert mix columns to string type
df_test['mix'] = df_test['mix'].astype('str')

## missing values

In [24]:
df_pro.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [27]:
#create subset of dataframe that contain null values
df_null = df_pro[df_pro['product_name'].isnull() == True]

In [28]:
df_null

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

## Task 

## Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [30]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_the_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06            3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01            2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01            2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00            0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00            1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01            3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01            5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02            6.000000e+00   

       order_hour_of_day  days_since_last_order  
count       3.421083e+06           3.214874e+06  
mean        1.345202e+01           1.111484e+01  
std         4.226088e+00           9.206737e+00  
min         0.000000e+00           0.000000e+00  
25%         1.000000e+01           4.000000e+00  
50%         1.300000e+01           7.000000e+00  
75%         1.600000e+01           1.500000e+01  
max         2.300000e+01           3.000000e+01

## step 3

In [34]:
## checking for mix values
for col in df_ords.columns.tolist():
    werid = (df_ords[col].map(type) != df_ords[col].iloc[0].__class__).any()
    if werid:
        print(col)

order_id
user_id
order_number
orders_day_of_the_week
order_hour_of_day
days_since_last_order


In [35]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_the_week  int64  
 4   order_hour_of_day       int64  
 5   days_since_last_order   float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


## there is no mix data in columns

## step 5

In [38]:
##checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_the_week         0
order_hour_of_day              0
days_since_last_order     206209
dtype: int64

In [39]:
df_ords.shape

(3421083, 6)

## days_since_last_order has 206209 missing values. i am asuming this cloud be customer first orders.

In [41]:
# create dataframe with null values
df_order_null = df_ords[df_ords['days_since_last_order'].isnull()==True]

In [42]:
df_order_null

order_id  user_id  order_number  orders_day_of_the_week  \
0         2539329        1             1                       2   
11        2168274        2             1                       2   
26        1374495        3             1                       1   
39        3343014        4             1                       6   
45        2717275        5             1                       3   
...           ...      ...           ...                     ...   
3420930    969311   206205             1                       4   
3420934   3189322   206206             1                       3   
3421002   2166133   206207             1                       6   
3421019   2227043   206208             1                       1   
3421069   3154581   206209             1                       3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

In [44]:
df_order_id_1 = df_order_null[df_order_null['order_number'] == 1]

In [45]:
df_order_id_1

order_id  user_id  order_number  orders_day_of_the_week  \
0         2539329        1             1                       2   
11        2168274        2             1                       2   
26        1374495        3             1                       1   
39        3343014        4             1                       6   
45        2717275        5             1                       3   
...           ...      ...           ...                     ...   
3420930    969311   206205             1                       4   
3420934   3189322   206206             1                       3   
3421002   2166133   206207             1                       6   
3421019   2227043   206208             1                       1   
3421069   3154581   206209             1                       3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

## missing data from columns days_since_last_order those for customers who order item first time. we can safly replace null value with 0

In [47]:
df_ords['days_since_last_order'].fillna(0, inplace = True)

C:\Users\hardi\AppData\Local\Temp\ipykernel_11848\2204563672.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords['days_since_last_order'].fillna(0, inplace = True)


In [48]:
df_ords.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_the_week    0
order_hour_of_day         0
days_since_last_order     0
dtype: int64

## step 7

In [50]:
#checking for dupicated value

In [53]:
df_ords_dup = df_ords[df_ords.duplicated()]

In [54]:
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_the_week, order_hour_of_day, days_since_last_order]
Index: []

## there is no dupicated data 

In [57]:
df_pro.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [59]:
df_pro.dropna(subset=['product_name'],inplace=True)

In [60]:
df_pro.isnull().sum()

product_id       0
product_name     0
aisle_id         0
department_id    0
prices           0
dtype: int64

In [61]:
df_ords.to_csv(os.path.join(path,"Data", "prepare_data","clean_oder_data.csv"),index=False)

In [62]:
df_pro.to_csv(os.path.join(path,"Data", "prepare_data","clean_products_data.csv"),index=False)